### PR TITLE
feat: add ws proxy functionality

### LINKED ISSUE #87: feat: add websocket support
Add websocket connection support to RPC proxy

### PR DESCRIPTION
**Pull request by Rakowskiii**
# Description

Adding handler for websocket proxying `/ws` 
Currently only supporting infura, as pokt doesn't seem to support websockets.
Zksync is deprecating ws https://docs.zksync.io/api/sdk/js/providers/#create-websocket-provider-it-is-deprecated-and-will-be-removed-soon

Resolves #87 

## How Has This Been Tested?

Tests ran for each connection.

## Due Diligence

* [ ] Breaking change
* [ ] Requires a documentation update
* [x] Requires a e2e/integration test update

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -47,11 +48,19 @@ pub enum RpcError {
...
+    #[error("Invalid scheme used. Try http(s):// or ws(s)://")]
+    InvalidScheme,
+
+    #[error(transparent)]
+    AxumTungstenite(#[from] axum_tungstenite::Error),
 }
 
 impl IntoResponse for RpcError {
     fn into_response(self) -> axum::response::Response {
+        error!("{:?}", self);
```
Comment (xDarksome) [inline src/error.rs:61]: Is this a recommended way to log errors in `axum`? I tend to consider logging inside conversion functions to be a bad practice.

Is there some other place we could `tap_err` or something?

File: src/extractors/protocol.rs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+use {
+    crate::error::RpcError,
+    async_trait::async_trait,
+    axum::{extract::FromRequestParts, http::request::Parts},
+};
+
+#[derive(Debug)]
+pub enum Protocol {
+    HTTP,
```
Comment (xDarksome) [inline src/extractors/protocol.rs:9]: Reasonable warning here. Should be `Http`

File: src/providers/mod.rs [hunk 3]
```diff
@@ -63,7 +78,24 @@ pub trait RpcProvider: Send + Sync {
         query_params: RpcQueryParams,
         headers: hyper::http::HeaderMap,
         body: hyper::body::Bytes,
-    ) -> RpcResult<Response<Body>>;
+    ) -> RpcResult<Response>;
+
+    fn supports_caip_chainid(&self, chain_id: &str) -> bool;
```
Comment (xDarksome) [inline src/providers/mod.rs:83]: Consider extracting those fn's into a separate trait and making this trait a supertrait of `RpcProvider` and `RpcWsProvider`.

File: tests/functional/websocket/infura.rs [hunk 4]
```diff
@@ -0,0 +1,411 @@
+use {
+    crate::{context::ServerContext, JSONRPC_VERSION},
+    futures_util::{SinkExt, StreamExt},
+    test_context::test_context,
+};
+
+#[test_context(ServerContext)]
+#[tokio::test]
+async fn eip155_1_mainnet_infura(ctx: &mut ServerContext) {
```
Comment (xDarksome) [inline tests/functional/websocket/infura.rs:9]: It's usually not a good idea to generalize tests, but these are doing almost exactly the same thing.  We have distinct "ok" and "gone" checks here. 

I would recommend writing two separate functions for those cases, parametrize them with a couple of required arguments and then use them inside all of the specific test cases.

File: src/error.rs (hunk 1 above)
Comment (Rakowskiii) [reply src/error.rs:61]: I don't think this is recommended way. 
My thought process behind this was: tapping in each place separately may lead to some uncovered spots in the future. Logging it here take off the responsibility from future contributors, as they don't need to keep remembering about each and every error they return.

File: tests/functional/websocket/infura.rs (hunk 4 above)
Comment (Rakowskiii) [reply tests/functional/websocket/infura.rs:12]: Done.

File: src/extractors/protocol.rs (hunk 2 above)
Comment (Rakowskiii) [reply src/extractors/protocol.rs]: Done.

File: src/providers/mod.rs [hunk 5]
```diff
@@ -55,16 +70,27 @@ impl Display for ProviderKind {
...
         xpath: axum::extract::MatchedPath,
         query_params: RpcQueryParams,
         headers: hyper::http::HeaderMap,
         body: hyper::body::Bytes,
-    ) -> RpcResult<Response<Body>>;
+    ) -> RpcResult<Response>;
+}
+
+#[async_trait]
+pub trait RpcWsProvider: Send + Sync + Provider {
```
Comment (xDarksome) [inline src/providers/mod.rs:85]: Nitpick: `Send + Sync + 'static`  are conventionally being written at the end

File: src/error.rs (hunk 1 above)
Comment (xDarksome) [reply src/error.rs:61]: We should be able to do logging via some middleware abstraction of `axum`, like `layer` or something, but I don't have a better solution rn. So LGTM for now

**Review by xDarksome — APPROVED** [review]

File: src/error.rs (hunk 1 above)
Comment (Rakowskiii) [reply src/error.rs:61]: I'll drop this as issue, thanks

**Pull request merged**